From c0be3c863c8f7a3306c34fb6384da6f35fbb7ae0 Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Fri, 14 Jul 2023 09:00:40 +0300
Subject: [PATCH] [SW-151130] gaudi2: add initialization flag to uart module

The prints in arc0 will impact pcie link up timings
the uart module is common for arc0 and arc1 and trying
to print from both the arcs can have sync issues.
This patch adds initialized flag in uart which is
default set to false and will be set to true as
part of device_init_uart. The device_init_uart
init is done only for arc1. Hence all the prints
from arc0 will be ignored.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: Icffc45a133e0a586c201a1ac464fd55e734484d9
---

diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index aedaea9..07fce4c 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -1200,6 +1200,7 @@
 	uart.uart_state_reg = mmPSOC_UART_0_STATE;
 	uart.uart_ctrl_reg = mmPSOC_UART_0_CTRL;
 	uart.uart_bauddiv_reg = mmPSOC_UART_0_BAUDDIV;
+	uart.regs_initialized = true;
 }
 
 void device_init_pci(void)
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index acf5116..707561f 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -528,6 +528,7 @@
 	uart.uart_state_reg = D0_PSOC_UART_0_REG(mmUART_STATE);
 	uart.uart_ctrl_reg = D0_PSOC_UART_0_REG(mmUART_CTRL);
 	uart.uart_bauddiv_reg = D0_PSOC_UART_0_REG(mmUART_BAUDDIV);
+	uart.regs_initialized = true;
 }
 
 void device_init_pci(void)
diff --git a/include/device.h b/include/device.h
index dc036f7..f8302df 100644
--- a/include/device.h
+++ b/include/device.h
@@ -18,6 +18,7 @@
 	u32 uart_state_reg;
 	u32 uart_ctrl_reg;
 	u32 uart_bauddiv_reg;
+	u32 regs_initialized;
 };
 
 struct device_pci {
diff --git a/lib/common/uart.c b/lib/common/uart.c
index 81bc783..690117f 100644
--- a/lib/common/uart.c
+++ b/lib/common/uart.c
@@ -18,7 +18,9 @@
 #define BITS_IN_SINGLE_CHAR		4
 #define RING_BUFF_SIZE			SZ_256
 
-struct device_uart uart;
+struct device_uart uart = {
+	.regs_initialized = false,
+};
 
 struct ring_buf_t {
 	char buf[RING_BUFF_SIZE];
@@ -57,6 +59,9 @@
 	unsigned int tail = tx_buf.tail;
 	char ch;
 
+	if (!uart.regs_initialized)
+		return;
+
 	/* if buffer is empty do nothing */
 	if (tx_buf.head != tail) {
 		ch = tx_buf.buf[tail];
@@ -73,6 +78,9 @@
 
 void preboot_uart_rbuf_flush(void)
 {
+	if (!uart.regs_initialized)
+		return;
+
 	while (tx_buf.head != tx_buf.tail) {
 		while (regmap_readl(uart.uart_state_reg) & UART_SR_TXFULL)
 			;
@@ -104,6 +112,9 @@
 
 void preboot_uart_init(unsigned int val)
 {
+	if (!uart.regs_initialized)
+		return;
+
 	regmap_writel(0x0, uart.uart_ctrl_reg);
 	regmap_writel(val, uart.uart_bauddiv_reg);
 	regmap_writel(0x3, uart.uart_ctrl_reg);
@@ -111,6 +122,9 @@
 
 int preboot_uart_tstc(void)
 {
+	if (!uart.regs_initialized)
+		return 0;
+
 	if (!(regmap_readl(uart.uart_state_reg) & UART_SR_RXFULL))
 		return 0;
 
@@ -119,6 +133,9 @@
 
 int preboot_uart_getc(void)
 {
+	if (!uart.regs_initialized)
+		return -1;
+
 	if (!preboot_uart_tstc())
 		return -1;
 
@@ -127,6 +144,9 @@
 
 bool preboot_uart_tx_full(void)
 {
+	if (!uart.regs_initialized)
+		return true;
+
 	return !!(regmap_readl(uart.uart_state_reg) & UART_SR_TXFULL);
 }
 
@@ -146,6 +166,9 @@
 
 void preboot_putc_raw(const char ch)
 {
+	if (!uart.regs_initialized)
+		return;
+
 	if (tx_buf.initialized) {
 		preboot_uart_rbuf_push(ch);
 		return;
@@ -161,12 +184,18 @@
 {
 	int i;
 
+	if (!uart.regs_initialized)
+		return;
+
 	for (i = 0; str[i] && (i < 128); i++)
 		preboot_putc_raw(str[i]);
 }
 
 void preboot_puts(const char *str)
 {
+	if (!uart.regs_initialized)
+		return;
+
 	preboot_puts_raw(str);
 	preboot_puts_raw("\r");
 }
@@ -177,6 +206,9 @@
 	char ch;
 	int shift_start = (STR_UINT_SZ - 1) * BITS_IN_SINGLE_CHAR;
 
+	if (!uart.regs_initialized)
+		return 0;
+
 	for (int i = shift_start; i >= 0; i -= BITS_IN_SINGLE_CHAR) {
 		ch = (num >> i) & 0xf;
 		/*
